internal/controllers/machine/machine_controller_inplace_update.go

```diff
@@ -0,0 +1,240 @@
+	if !hasUpdateMachinePending {
+		log.Info("In-place update annotations are set, waiting for UpdateMachine hook to be marked as pending")
+		return ctrl.Result{}, nil
+	}
+
+	if s.infraMachine == nil {
+		s.updatingReason = clusterv1.MachineUpdateFailedReason
+		s.updatingMessage = "In-place update not possible: InfraMachine not found"
+		return ctrl.Result{}, errors.New("in-place update failed: InfraMachine not found")
+	}
+
+	infraReady := r.isInfraMachineReadyForUpdate(s)
+	bootstrapReady := r.isBootstrapConfigReadyForUpdate(s)
+
+	if !infraReady || !bootstrapReady {
+		log.Info("Waiting for InfraMachine and BootstrapConfig to be marked for in-place update")
+		return ctrl.Result{}, nil
+	}
+
+	if hasUpdateMachinePending {
+		log.Info("UpdateMachine hook is pending, calling runtime hook")
+		result, err := r.callUpdateMachineHook(ctx, s)
+		if err != nil {
+			s.updatingReason = clusterv1.MachineUpdateFailedReason
+			s.updatingMessage = fmt.Sprintf("UpdateMachine hook failed: %v", err)
+			return ctrl.Result{}, err
+		}
+
+		if result.RequeueAfter > 0 {
+			s.updatingReason = clusterv1.MachineWaitingForUpdateMachineHookReason
+			s.updatingMessage = "UpdateMachine hook in progress"
+			return result, nil
+		}
+
+		log.Info("In-place update completed successfully")
+		if err := r.completeInPlaceUpdate(ctx, s); err != nil {
+			return ctrl.Result{}, errors.Wrap(err, "failed to complete in-place update")
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// isInfraMachineReadyForUpdate checks if the InfraMachine has the in-place update annotation.
+func (r *Reconciler) isInfraMachineReadyForUpdate(s *scope) bool {
+	_, hasAnnotation := s.infraMachine.GetAnnotations()[clusterv1.InPlaceUpdateInProgressAnnotation]
+	return hasAnnotation
+}
+
+// isBootstrapConfigReadyForUpdate checks if the BootstrapConfig has the in-place update annotation.
+func (r *Reconciler) isBootstrapConfigReadyForUpdate(s *scope) bool {
+	if s.bootstrapConfig == nil {
+		return true
+	}
+	_, hasAnnotation := s.bootstrapConfig.GetAnnotations()[clusterv1.InPlaceUpdateInProgressAnnotation]
+	return hasAnnotation
+}
+
+// callUpdateMachineHook calls the UpdateMachine runtime hook for the machine.
+func (r *Reconciler) callUpdateMachineHook(ctx context.Context, s *scope) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	request := &runtimehooksv1.UpdateMachineRequest{
+		Desired: runtimehooksv1.UpdateMachineRequestObjects{
+			Machine:               *cleanupMachine(s.machine),
+			InfrastructureMachine: runtime.RawExtension{Object: cleanupUnstructured(s.infraMachine)},
+		},
+	}
+
+	if s.bootstrapConfig != nil {
+		request.Desired.BootstrapConfig = runtime.RawExtension{Object: cleanupUnstructured(s.bootstrapConfig)}
+	}
+
+	response := &runtimehooksv1.UpdateMachineResponse{}
+
+	if err := r.RuntimeClient.CallAllExtensions(ctx, runtimehooksv1.UpdateMachine, s.machine, request, response); err != nil {
```

Comment:
Once #12857 merges.

Can you please call GetAllExtensions first and check if we have any extensions.
* fail if there are no extensions registered
* fail if there are > 1 extensions registered (for now for the current iteration, see: https://github.com/kubernetes-sigs/cluster-api/pull/12857#discussion_r2440966038)
  * We want to introduce support for multiple extensions in a safe way later

(cc @fabriziopandini)